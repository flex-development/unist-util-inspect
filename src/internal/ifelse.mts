/**
 * @file Internal - ifelse
 * @module unist-util-inspect/internal/ifelse
 */

export default ifelse

/**
 * Return `truthy` if `condition` is `true`, or `falsy` otherwise.
 *
 * @internal
 *
 * @template T
 *  Truthy value type
 * @template F
 *  Falsy value type
 *
 * @this {void}
 *
 * @param {boolean | ((this: void) => boolean)} condition
 *  The condition to evaluate
 * @param {T} truthy
 *  The value to return if `condition` evaluates to `true`
 * @param {F} falsy
 *  The value to return if `condition` evaluates to `false`
 * @return {F | T}
 *  Condition result
 */
function ifelse<T, F>(
  this: void,
  condition: boolean | ((this: void) => boolean),
  truthy: T,
  falsy: F
): F | T {
  if (typeof condition !== 'boolean') condition = condition()
  return condition ? truthy : falsy
}
