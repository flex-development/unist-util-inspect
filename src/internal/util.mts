/**
 * @file Internal - util
 * @module unist-util-inspect/internal/util
 */

import type State from '#types/state'
import {
  stringifyPosition
} from '@flex-development/unist-util-stringify-position'
import type { Literal, Node, Parent } from 'unist'

export { inspectValue }

/**
 * Bold the given `value`.
 *
 * @internal
 *
 * @param {string} value
 *  The value to color
 * @return {string}
 *  Colored `value`
 */
const bold: (value: string) => string = ansi(1, 22)

/**
 * Dim the given `value`.
 *
 * @internal
 *
 * @param {string} value
 *  The value to color
 * @return {string}
 *  Colored `value`
 */
const dim: (value: string) => string = ansi(2, 22)

/**
 * Color the given `value` green.
 *
 * @internal
 *
 * @param {string} value
 *  The value to color
 * @return {string}
 *  Colored `value`
 */
const green: (value: string) => string = ansi(32, 39)

/**
 * Color the given `value` yellow.
 *
 * @internal
 *
 * @param {string} value
 *  The value to color
 * @return {string}
 *  Colored `value`
 */
const yellow: (value: string) => string = ansi(33, 39)

/**
 * Factory to wrap values in ANSI colors.
 *
 * @internal
 *
 * @param {number} open
 *  The opening color code
 * @param {number} close
 *  The closing color code
 * @return {(value: string) => string}
 *  Function to colorize a value
 */
function ansi(open: number, close: number): (value: string) => string {
  return color

  /**
   * Colorize `value`.
   *
   * @param {string} value
   *  The value to color
   * @return {string}
   *  Colored `value`
   */
  function color(value: string): string {
    return '\u001B[' + open + 'm' + value + '\u001B[' + close + 'm'
  }
}

/**
 * Indent `value`.
 *
 * @internal
 *
 * @param {string} value
 *  The value to ident
 * @param {string} indentation
 *  The indent to use
 * @param {boolean | undefined} [skipFirst]
 *  Whether to skip indenting first line
 * @return {string}
 *  Indented `value`
 */
function indent(
  value: string,
  indentation: string,
  skipFirst?: boolean | undefined
): string {
  if (!value) return value

  /**
   * Lines to indent.
   *
   * @const {string[]} lines
   */
  const lines: string[] = value.split('\n')

  /**
   * Index of current line.
   *
   * @var {number} index
   */
  let index: number = skipFirst ? 0 : -1

  // indent lines
  while (++index < lines.length) lines[index] = indentation + lines[index]!

  return lines.join('\n')
}

/**
 * Format the fields of a node.
 *
 * @internal
 *
 * @param {Record<string, any>} fields
 *  Fields to format
 * @param {State} state
 *  Info passed around
 * @return {string}
 *  Pretty printed `fields`
 */
function inspectFields(fields: Record<string, any>, state: State): string {
  /**
   * Formatted parts array.
   *
   * @const {string[]} parts
   */
  const parts: string[] = []

  /**
   * Current key.
   *
   * @var {string} key
   */
  let key: string

  for (key in fields) {
    if (Object.prototype.hasOwnProperty.call(fields, key)) {
      /**
       * Field value.
       *
       * @const {any} value
       */
      const value: any = fields[key]

      if (
        // standard keys defined by unist that are formatted differently
        // <https://github.com/syntax-tree/unist>
        key === 'children' ||
        key === 'position' ||
        key === 'type' ||
        key === 'value' ||
        // ignore `name` (esast, xast) and `tagName` (`hast`)
        (typeof value === 'string' && (key === 'name' || key === 'tagName'))
      ) {
        continue
      }

      /**
       * Formatted value.
       *
       * @var {string} formatted
       */
      let formatted: string

      if (
        nodelike(value) &&
        key !== 'attributes' &&
        key !== 'data' &&
        key !== 'properties'
      ) {
        formatted = inspectTree(value, state)
      } else if (Array.isArray(value) && nodelike(value[0])) {
        formatted = '\n' + inspectNodes(value, state)
      } else {
        formatted = inspectNonTree(value)
      }

      parts.push(
        key + dim(':') + (/\s/.test(formatted.charAt(0)) ? '' : ' ') + formatted
      )
    }
  }

  return indent(
    parts.join('\n'),
    (Array.isArray(fields['children']) && fields['children'].length > 0
      ? dim('│')
      : ' ') + ' '
  )
}

/**
 * Format a node.
 *
 * @internal
 *
 * @param {Node | Literal | Parent} node
 *  The node to format
 * @param {State} state
 *  Info passed around
 * @return {string}
 *  Pretty printed `node`
 */
function inspectNode(node: Node | Literal | Parent, state: State): string {
  /**
   * Formatted parts array.
   *
   * @const {string[]} parts
   */
  const parts: string[] = [bold(node.type)]

  /**
   * Node kind.
   *
   * @const {string | undefined} kind
   */
  const kind: string | undefined =
    'tagName' in node && typeof node.tagName === 'string'
      ? node.tagName
      : 'name' in node && typeof node.name === 'string'
      ? node.name
      : undefined

  // format node kind
  if (kind) parts.push('<' + kind + '>')

  // format node children or value
  if ('children' in node && Array.isArray(node.children)) {
    parts.push(dim('['), yellow(String(node.children.length)), dim(']'))
  } else if ('value' in node) {
    if (
      typeof node.value !== 'function' &&
      (typeof node.value !== 'object' || node.value === null) &&
      typeof node.value !== 'symbol'
    ) {
      parts.push(' ', green(inspectNonTree(node.value)))
    }
  }

  // format position
  if (state.positions && node.position) {
    parts.push(' ', dim('('))
    parts.push(stringifyPosition(node, { offsets: true }), dim(')'))
  }

  return parts.join('')
}

/**
 * Format a list of nodes.
 *
 * @internal
 *
 * @param {unknown[]} nodes
 *  The nodes to format
 * @param {State} state
 *  Info passed around
 * @return {string}
 *  Pretty printed `nodes`
 */
function inspectNodes(nodes: unknown[], state: State): string {
  /**
   * List size.
   *
   * @const {number} size
   */
  const size: number = String(nodes.length - 1).length

  /**
   * Formatted parts array.
   *
   * @const {string[]} parts
   */
  const parts: string[] = []

  /**
   * Index of current node.
   *
   * @var {number} index
   */
  let i: number = -1

  // format node list
  while (++i < nodes.length) {
    /**
     * Last node?
     *
     * @const {boolean} end
     */
    const end: boolean = i === nodes.length - 1

    /**
     * Formatted part.
     *
     * @var {string} part
     */
    let part: string = dim((end ? '└' : '├') + '─' + String(i).padEnd(size))

    // finish formatting
    part += ' ' + indent(
      inspectValue(nodes[i], state),
      (end ? ' ' : dim('│')) + ' '.repeat(size + 2),
      true
    )

    parts.push(part)
  }

  return parts.join('\n')
}

/**
 * Format an unknown value.
 *
 * @internal
 *
 * @param {unknown} value
 *  The thing to format
 * @return {string}
 *  Pretty printed `value`
 */
function inspectNonTree(value: unknown): string {
  return value === undefined
    ? 'undefined'
    : typeof value === 'bigint'
    ? value.toLocaleString() + 'n'
    : JSON.stringify(value)
}

/**
 * Format a `node`, its fields, and its children.
 *
 * @internal
 *
 * @param {Node | Literal | Parent} node
 *  The node to format
 * @param {State} state
 *  Info passed around
 * @return {string}
 *  Pretty printed `node`
 */
function inspectTree(node: Node | Literal | Parent, state: State): string {
  /**
   * Formatted parts array.
   *
   * @const {string[]} parts
   */
  const parts: string[] = [inspectNode(node, state)]

  /**
   * Formatted fields.
   *
   * @const {string} fields
   */
  const fields: string = inspectFields(node, state)

  /**
   * Formatted node children.
   *
   * @const {string} children
   */
  const children: string = 'children' in node && Array.isArray(node.children)
    ? inspectNodes(node.children, state)
    : ''

  if (fields) parts.push(fields)
  if (children) parts.push(children)

  return parts.join('\n')
}

/**
 * Format any value.
 *
 * @internal
 *
 * @param {unknown} value
 *  The thing to format
 * @param {State} state
 *  Info passed around
 * @return {string}
 *  Pretty printed `value`
 */
function inspectValue(value: unknown, state: State): string {
  if (Array.isArray(value)) return inspectNodes(value, state)
  if (nodelike(value)) return inspectTree(value, state)
  return inspectNonTree(value)
}

/**
 * Check if something looks like a {@linkcode Node}.
 *
 * @internal
 *
 * @param {unknown} value
 *  The thing to check
 * @return {value is Node}
 *  `true` if `value` looks like a node, `false` otherwise
 */
function nodelike(value: unknown): value is Node {
  return (
    typeof value === 'object' &&
    value !== null &&
    'type' in value &&
    typeof value.type === 'string'
  )
}
