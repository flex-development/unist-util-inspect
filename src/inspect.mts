/**
 * @file inspect
 * @module unist-util-inspect/inspect
 */

import ifelse from '#internal/ifelse'
import { inspectValue } from '#internal/util'
import { isColorSupported, stripAnsi } from '@flex-development/colors'
import type { Inspect, Options } from '@flex-development/unist-util-inspect'

export { inspect, inspectColor, inspectNoColor }

/**
 * Inspect a tree with color in supported environments or without color in
 * environments that do not support color.
 *
 * @see {@linkcode Inspect}
 *
 * @const {Inspect} inspect
 */
const inspect: Inspect = ifelse(isColorSupported, inspectColor, inspectNoColor)

/**
 * Inspect a `tree` with color.
 *
 * @see {@linkcode Options}
 *
 * @this {void}
 *
 * @param {unknown} tree
 *  The tree to inspect
 * @param {Options | null | undefined} [options]
 *  Configuration options
 * @return {string}
 *  Pretty printed `tree`
 */
function inspectColor(
  this: void,
  tree: unknown,
  options?: Options | null | undefined
): string {
  return inspectValue(tree, { positions: options?.positions ?? true })
}

/**
 * Inspect a `tree` without color.
 *
 * @see {@linkcode Options}
 *
 * @this {void}
 *
 * @param {unknown} tree
 *  The tree to inspect
 * @param {Options | null | undefined} [options]
 *  Configuration options
 * @return {string}
 *  Pretty printed `tree`
 */
function inspectNoColor(
  this: void,
  tree: unknown,
  options?: Options | null | undefined
): string {
  return stripAnsi(inspectColor(tree, options))
}
