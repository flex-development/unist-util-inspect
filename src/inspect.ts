/**
 * @file inspect
 * @module unist-util-inspect/inspect
 */

import { color } from '#conditional-color'
import type { Options } from './types'
import { inspectValue } from './util'

/**
 * Regular expression matching an ANSI color.
 *
 * @const {RegExp} COLOR_RE
 */
const COLOR_RE: RegExp =
  /(?:(?:\u001B\[)|\u009B)(?:\d{1,3})?(?:(?:;\d{0,3})*)?[A-Mf-m|]|\u001B[A-M]/g

/**
 * Inspect a node, with color in Node, without color in browsers.
 *
 * @see {@linkcode Options}
 *
 * @param {unknown} tree - Tree to inspect
 * @param {(Options | null)?} [options] - Configuration options
 * @return {string} Pretty printed `tree`
 */
/* c8 ignore next 3 */
const inspect: (tree: unknown, options?: Options | null) => string = color
  ? inspectColor
  : inspectNoColor

/**
 * Inspect a node, using color.
 *
 * @see {@linkcode Options}
 *
 * @param {unknown} tree - Tree to inspect
 * @param {(Options | null)?} [options] - Configuration options
 * @return {string} Pretty printed `tree`
 */
function inspectColor(tree: unknown, options?: Options | null): string {
  return inspectValue(tree, { positions: options?.positions ?? true })
}

/**
 * Inspect a node, without color.
 *
 * @see {@linkcode Options}
 *
 * @param {unknown} tree - Tree to inspect
 * @param {(Options | null)?} [options] - Configuration options
 * @return {string} Pretty printed `tree`
 */
function inspectNoColor(tree: unknown, options?: Options | null): string {
  return inspectColor(tree, options).replace(COLOR_RE, '')
}

export { inspect, inspectColor, inspectNoColor }
