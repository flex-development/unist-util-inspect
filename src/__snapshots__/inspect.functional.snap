// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`functional:inspect > should handle nodes outside of children (sample 0) 1`] = `
"element<template>
  content: root[0]"
`;

exports[`functional:inspect > should handle nodes outside of children (sample 1) 1`] = `
"jsxSpan[1]
│ close: jsxTag
│   attributes: []
│   close: true
│   name: jsxMember
│     object: jsxMember
│       object: jsxIdentifier "abc"
│       property: jsxIdentifier "def"
│     property: jsxIdentifier "ghi"
│   selfClosing: false
│ open: jsxTag
│   attributes:
│   ├─0 jsxAttribute null
│   │     name: jsxIdentifier "alpha"
│   ├─1 jsxAttributeExpression "...props"
│   └─2 jsxAttribute null
│         name: jsxIdentifier "bravo"
│   close: false
│   name: jsxMember
│     object: jsxMember
│       object: jsxIdentifier "abc"
│       property: jsxIdentifier "def"
│     property: jsxIdentifier "ghi"
│   selfClosing: false
└─0 jsxExpressionSpan "1 + 1""
`;

exports[`functional:inspect > should see attributes as data 1`] = `
"root[1] (1:1-1:19, 0-18)
└─0 element<album>[0] (1:1-1:19, 0-18)
      attributes: {"id":"123"}"
`;

exports[`functional:inspect > should see data as data 1`] = `
"album
  data: {"id":"123"}"
`;

exports[`functional:inspect > should see properties as data 1`] = `
"element<button>[0]
  properties: {"type":"submit","value":"Send"}"
`;

exports[`functional:inspect > should support \`positions: false\` 1`] = `
"comment[3]
├─0 blockTag<@file>[1]
│   └─0 text "Fixtures - average"
├─1 blockTag<@module>[1]
│   └─0 text "fixtures/average"
└─2 blockTag<@see>[1]
    └─0 text "https://codewars.com/kata/57a2013acf1fa5bfc4000921""
`;

exports[`functional:inspect > should work with list of nodes (sample 0) 1`] = `""`;

exports[`functional:inspect > should work with list of nodes (sample 1) 1`] = `
"├─0 comment[3] (1:1-5:4, 0-116)
│   ├─0 blockTag<@file>[1] (2:4-2:25, 7-28)
│   │   └─0 text "Fixtures - diff" (2:10-2:25, 13-28)
│   ├─1 blockTag<@module>[1] (3:4-3:25, 32-53)
│   │   └─0 text "fixtures/diff" (3:12-3:25, 40-53)
│   └─2 blockTag<@see>[1] (4:4-4:59, 57-112)
│       └─0 text "https://codewars.com/kata/523f5d21c841566fde000009" (4:9-4:59, 62-112)
└─1 comment[8] (7:1-23:4, 118-582)
    ├─0 description[1] (8:4-9:37, 125-237)
    │   └─0 paragraph[9] (8:4-9:37, 125-237)
    │       ├─0 text "Given two number arrays, " (8:4-8:29, 125-150)
    │       ├─1 inlineCode "a" (8:29-8:32, 150-153)
    │       ├─2 text " and " (8:32-8:37, 153-158)
    │       ├─3 inlineCode "b" (8:37-8:40, 158-161)
    │       ├─4 text ", the function returns a new array with elements in " (8:40-9:16, 161-216)
    │       ├─5 inlineCode "b" (9:16-9:19, 216-219)
    │       ├─6 text " removed from " (9:19-9:33, 219-233)
    │       ├─7 inlineCode "a" (9:33-9:36, 233-236)
    │       └─8 text "." (9:36-9:37, 236-237)
    ├─1 blockTag<@example>[1] (11:4-12:27, 244-279)
    │   └─0 code "diff([], [4, 5]) // []" (12:5-12:27, 257-279)
    │         lang: null
    │         meta: null
    ├─2 blockTag<@example>[1] (13:4-14:29, 283-320)
    │   └─0 code "diff([3, 4], [3]) // [4]" (14:5-14:29, 296-320)
    │         lang: null
    │         meta: null
    ├─3 blockTag<@example>[1] (15:4-16:35, 324-367)
    │   └─0 code "diff([1, 2, 3], [1, 2]) // [3]" (16:5-16:35, 337-367)
    │         lang: null
    │         meta: null
    ├─4 blockTag<@example>[1] (17:4-18:37, 371-416)
    │   └─0 code "diff([1, 8, 2], []) // [1, 8, 2]" (18:5-18:37, 384-416)
    │         lang: null
    │         meta: null
    ├─5 blockTag<@param>[2] (20:4-20:44, 423-463)
    │   ├─0 typeExpression "number[]" (20:11-20:21, 430-440)
    │   └─1 text "a - First number array" (20:22-20:44, 441-463)
    ├─6 blockTag<@param>[2] (21:4-21:45, 467-508)
    │   ├─0 typeExpression "number[]" (21:11-21:21, 474-484)
    │   └─1 text "b - Second number array" (21:22-21:45, 485-508)
    └─7 blockTag<@return>[5] (22:4-22:70, 512-578)
        ├─0 typeExpression "number[]" (22:12-22:22, 520-530)
        ├─1 text "New array with elements in " (22:23-22:50, 531-558)
        ├─2 inlineCode "b" (22:50-22:53, 558-561)
        ├─3 text " removed from " (22:53-22:67, 561-575)
        └─4 inlineCode "a" (22:67-22:70, 575-578)"
`;

exports[`functional:inspect > should work with literals (sample 0) 1`] = `"bigint 13n"`;

exports[`functional:inspect > should work with literals (sample 1) 1`] = `"boolean true"`;

exports[`functional:inspect > should work with literals (sample 2) 1`] = `"number 3"`;

exports[`functional:inspect > should work with literals (sample 3) 1`] = `
"regexp "/(?<=(?<a>\\\\w){3})f/u"
  flags: "u"
  pattern: "(?<=(?<a>\\\\w){3})f""
`;

exports[`functional:inspect > should work with literals (sample 4) 1`] = `"string """`;

exports[`functional:inspect > should work with literals (sample 5) 1`] = `"null null"`;

exports[`functional:inspect > should work with literals (sample 6) 1`] = `"undefined undefined"`;

exports[`functional:inspect > should work with non-nodes (sample 0) 1`] = `""node""`;

exports[`functional:inspect > should work with non-nodes (sample 1) 1`] = `"13"`;

exports[`functional:inspect > should work with non-nodes (sample 2) 1`] = `"null"`;

exports[`functional:inspect > should work with non-nodes (sample 3) 1`] = `"null"`;

exports[`functional:inspect > should work with parents (sample 0) 1`] = `"root[0]"`;

exports[`functional:inspect > should work with parents (sample 1) 1`] = `
"table[2]
│ align: ["center","left"]
├─0 tableRow[2]
│   ├─0 tableCell[1]
│   │   └─0 text "foo"
│   └─1 tableCell[1]
│       └─0 text "bar"
└─1 tableRow[2]
    ├─0 tableCell[1]
    │   └─0 text "baz"
    └─1 tableCell[1]
        └─0 text "qux""
`;

exports[`functional:inspect > should work with parents (sample 2) 1`] = `
"root[4]
├─0 comment[3] (1:1-5:4, 0-114)
│   ├─0 blockTag<@file>[1] (2:4-2:24, 7-27)
│   │   └─0 text "Fixtures - hrt" (2:10-2:24, 13-27)
│   ├─1 blockTag<@module>[1] (3:4-3:24, 31-51)
│   │   └─0 text "fixtures/hrt" (3:12-3:24, 39-51)
│   └─2 blockTag<@see>[1] (4:4-4:59, 55-110)
│       └─0 text "https://codewars.com/kata/52685f7382004e774f0001f7" (4:9-4:59, 60-110)
├─1 comment[6] (7:1-20:4, 116-473)
│   ├─0 description[1] (8:4-9:39, 123-238)
│   │   └─0 paragraph[5] (8:4-9:39, 123-238)
│   │       ├─0 text "Given a non-negative integer, " (8:4-8:34, 123-153)
│   │       ├─1 inlineCode "seconds" (8:34-8:43, 153-162)
│   │       ├─2 text ", the functions returns the time in a human-readable format (" (8:43-9:27, 162-226)
│   │       ├─3 inlineCode "HH:MM:SS" (9:27-9:37, 226-236)
│   │       └─4 text ")." (9:37-9:39, 236-238)
│   ├─1 blockTag<@example>[1] (11:4-12:25, 245-278)
│   │   └─0 code "hrt(0) // '00:00:00'" (12:5-12:25, 258-278)
│   │         lang: null
│   │         meta: null
│   ├─2 blockTag<@example>[1] (13:4-14:26, 282-316)
│   │   └─0 code "hrt(60) // '00:01:00'" (14:5-14:26, 295-316)
│   │         lang: null
│   │         meta: null
│   ├─3 blockTag<@example>[1] (15:4-16:30, 320-358)
│   │   └─0 code "hrt(359999) // '99:59:59'" (16:5-16:30, 333-358)
│   │         lang: null
│   │         meta: null
│   ├─4 blockTag<@param>[2] (18:4-18:45, 365-406)
│   │   ├─0 typeExpression "number" (18:11-18:19, 372-380)
│   │   └─1 text "seconds - Time in seconds" (18:20-18:45, 381-406)
│   └─5 blockTag<@return>[4] (19:4-19:63, 410-469)
│       ├─0 typeExpression "string" (19:12-19:20, 418-426)
│       ├─1 text "Time in human-readable format (" (19:21-19:52, 427-458)
│       ├─2 inlineCode "HH:MM:SS" (19:52-19:62, 458-468)
│       └─3 text ")" (19:62-19:63, 468-469)
├─2 comment[2] (22:3-26:6, 519-613)
│   ├─0 description[1] (23:6-23:51, 528-573)
│   │   └─0 paragraph[2] (23:6-23:51, 528-573)
│   │       ├─0 inlineTag<@linkcode> "seconds" (23:6-23:25, 528-547)
│   │       └─1 text " in human-readable format." (23:25-23:51, 547-573)
│   └─1 blockTag<@var>[2] (25:6-25:29, 584-607)
│       ├─0 typeExpression "string" (25:11-25:19, 589-597)
│       └─1 text "formatted" (25:20-25:29, 598-607)
└─3 comment[2] (31:5-35:8, 742-846)
    ├─0 description[1] (32:8-32:58, 753-803)
    │   └─0 paragraph[2] (32:8-32:58, 753-803)
    │       ├─0 inlineTag<@linkcode> "seconds" (32:8-32:27, 753-772)
    │       └─1 text " in hours, minutes, or seconds." (32:27-32:58, 772-803)
    └─1 blockTag<@const>[2] (34:8-34:28, 818-838)
        ├─0 typeExpression "number" (34:15-34:23, 825-833)
        └─1 text "time" (34:24-34:28, 834-838)"
`;

exports[`functional:inspect > should work with void nodes 1`] = `"break"`;
